rpk: add is_alive info in 'redpanda admin broker ls'

daisukebe opened the pull request:
## Cover letter

Let `rpk redpanda admin brokers ls` return the broker's health status.

As-is
```
NODE ID  NUM CORES  MEMBERSHIP STATUS
1        16         active
2        16         active
3        16         active
```

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes https://github.com/redpanda-data/redpanda/issues/3998

## Release notes
* none

### Improvements

To-be
```
NODE_ID  NUM_CORES  MEMBERSHIP_STATUS  IS_ALIVE
1        16         active             true
2        16         active             true
3        16         active             true
```

daisukebe commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -54,10 +54,10 @@ func newListCommand(fs afero.Fs) *cobra.Command {
 			out.MaybeDie(err, "unable to request brokers: %v", err)
 
-			tw := out.NewTable("Node ID", "Num Cores", "Membership Status")
+			tw := out.NewTable("Node_ID", "Num_Cores", "Membership_Status", "Is_Alive")
```
It's based upon the user's feedback, `All the status fields are empty`. I wanted to make what the column name is explicit. I'm happy to see how you feel it.

twmb replied:
My only preference is dashes for consistency with other commands, but, one benefit of non-spaces (even though I introduced spaces) is better awking.

jcsp commented:
I'm not sure it matters, but be aware that the is_alive attribute was only added in 21.11.1 -- this won't work properly if someone tries to use this rpk against an older redpanda.  @0x5d did we ever define any rules about how far back rpk is supposed to be compatible?

twmb commented:
We can hide the field if it is missing by switching `IsAlive` to a pointer: if the value is `nil` (unspecified, default zero value is nil pointer) then we can elide the column.

Perhaps we should start returning the redpanda version in all http responses to avoid these forward/backward compatibility problems.

0x5d commented:
@jcsp good catch, thanks for the context. I agree with @twmb in both things he suggested.

daisukebe commented:
Great catch and thanks for the advise @jcsp and @twmb !

Do we expect something like this when the `Is_Alive` field doesn't exist?

```
NODE_ID  NUM_CORES  MEMBERSHIP_STATUS  IS_ALIVE
1        16         active             <nil>
2        16         active             <nil>
3        16         active             <nil>
```

Otherwise should I remove the `IS ALIVE` column itself when it doesn't exist?

twmb commented:
I think something like this (conceptually): https://github.com/redpanda-data/redpanda/blob/3736fe50a33a309c645b5f225a061fd4f9b536ef/src/go/rpk/pkg/cli/cmd/group/describe.go#L179-L184

So, if any of the IsAlive pointers are non-nil, then add the column; for any that are still know, print something like `unknown`. If all pointers are nil, elide the column.

Also, I think dashes are more standard in the command in comparison to underscores -- `IS-ALIVE` vs `IS_ALIVE` (though spaces might work here too).

daisukebe commented:
Upon playing around a bit, here's what I've figured out.

* The admin client tries to connect one available node among the list randomly
* If the reached node has the `is_alive` capability, all `Broker` instances have the `IsAlive` flag. 
    * If the node **doesn't** have it, all `Broker` don't have the flag

Given this,

* if any of the IsAlive pointers are non-nil, then add the column --> yes
* If all pointers are nil, elide the column. --> yes
* for any that are still nil, print something like unknown. --> this doesn't seem to happen when any of the `IsAlive` pointers are non-nil

Am I understanding properly? @twmb 

My new patch is supposed to show up the column when the reached node has the capability like below

Node-Id 1 --> 0.0.0.0:9644 running the shipped 21.11.8
Node-Id 2 --> 0.0.0.0:9645 running a customized 21.11.8 which I've removed the Is_Alive capability. It's stopped in this test
Node-Id 3 --> 0.0.0.0:9646 running a customized 21.11.8 which I've removed the Is_Alive capability
```
$ devrpk redpanda admin brokers --hosts 0.0.0.0:9644 ls
NODE-ID  NUM-CORES  MEMBERSHIP-STATUS  IS-ALIVE
1        16         active             true
2        16         active             false
3        16         active             true
$
$ devrpk redpanda admin brokers --hosts 0.0.0.0:9646 ls
NODE-ID  NUM-CORES  MEMBERSHIP-STATUS
1        16         active
2        16         active
3        16         active
```

What do you think?

twmb reviewed (dismissed):
Awesome yep, that looks great to me, thank you!

daisukebe commented:
Thanks for reviewing. I have just merged the two commits down to a single commit.

daisukebe commented:
Just wondering if we need to wait for the other reviewers' 👍 before committing to dev.

jcsp commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -54,10 +54,26 @@ func newListCommand(fs afero.Fs) *cobra.Command {
+			}
+			for _, b := range bs {
+				if b.IsAlive != nil {
+					headers = append(headers, "Is-Alive")
```
We're inside a loop over the brokers so isn't this going to append an extra Is-Alive column for each broker, instead of just one column that is populated for each broker?

twmb replied:
The break just below quits this loop, the bit of logic in here is to ensure the column exists and to change the args func to account for the new column

jcsp replied:
Silly me, that was a pre-coffee code review.  Never mind.

twmb commented:
/backport v22.1.x
